src/accountsdb/download.zig

```diff
@@ -538,10 +544,171 @@ pub fn downloadFile(
     }
 }
 
-const LegacyContactInfo = sig.gossip.data.LegacyContactInfo;
-const SignedGossipData = sig.gossip.data.SignedGossipData;
+pub fn getOrDownloadAndUnpackSnapshot(
+    allocator: std.mem.Allocator,
+    logger_: Logger,
+    validator_dir: std.fs.Dir,
+    options: struct {
+        /// gossip service is not needed when loading from an existing snapshot.
+        /// but when we need to download a new snapshot (force_new_snapshot_download flag),
+        /// we need the gossip service.
+        gossip_service: ?*GossipService = null,
+        force_new_snapshot_download: bool = false,
+        force_unpack_snapshot: bool = false,
+        num_threads_snapshot_unpack: u16 = 0,
+        min_snapshot_download_speed_mbs: usize = 20,
+        trusted_validators: ?[]const Pubkey = null,
+    },
+) !struct { AllSnapshotFields, SnapshotFiles } {
+    const logger = logger_.withScope(LOG_SCOPE);
 
-const KeyPair = std.crypto.sign.Ed25519.KeyPair;
+    const force_unpack_snapshot = options.force_unpack_snapshot;
+    const force_new_snapshot_download = options.force_new_snapshot_download;
+    var n_threads_snapshot_unpack: u32 = options.num_threads_snapshot_unpack;
+    if (n_threads_snapshot_unpack == 0) {
+        const n_cpus = @as(u32, @truncate(try std.Thread.getCpuCount()));
+        n_threads_snapshot_unpack = n_cpus / 2;
+    }
+
+    // check if we need to download a fresh snapshot
+    var accounts_db_exists = blk: {
+        if (validator_dir.openDir("accounts_db", .{ .iterate = true })) |dir| {
+            std.posix.close(dir.fd);
+            break :blk true;
+        } else |_| {
+            break :blk false;
+        }
+    };
+
+    // clear old snapshots, if we will download a new one
+    if (force_new_snapshot_download and accounts_db_exists) {
+        logger.info().log("deleting accounts_db dir...");
+        try validator_dir.deleteTreeMinStackSize("accounts_db");
+        accounts_db_exists = false;
+    }
+
+    if (!accounts_db_exists) try validator_dir.makeDir("accounts_db");
```

Comment:
Might be possible to replace `validator_dir.makeDir` and `validator_dir.openDir` with [makeOpenPath](https://ziglang.org/documentation/master/std/#std.fs.Dir.makeOpenPath)

Reply:
0d198119